eventlog/el/main.go

```diff
@@ -347,3 +411,327 @@ func search(c *cli.Context) error {
+}
+
+type openidCfg struct {
+	Issuer     string
+	Config     oauth2.Config
+	Token      oauth2.Token
+	Data       string
+	Public     kyber.Point
+	curRefresh string
+}
+
+func (o *openidCfg) save() (string, error) {
+	// Do not save when not needed.
+	if o.Token.RefreshToken == o.curRefresh {
+		return "", nil
+	}
+
+	dataDir := cfgpath.GetDataPath(cliApp.Name)
+	os.MkdirAll(dataDir, 0755)
+	fn := filepath.Join(dataDir, "openid.cfg")
+
+	// perms = 0600 because there is key material inside this file.
+	f, err := os.OpenFile(fn, os.O_RDWR|os.O_CREATE, 0600)
+	if err != nil {
+		return fn, fmt.Errorf("could not write %v: %v", fn, err)
+	}
+
+	buf, err := network.Marshal(o)
+	if err != nil {
+		return fn, err
+	}
+	_, err = f.Write(buf)
+	if err != nil {
+		return fn, err
+	}
+
+	// Remember the current refresh token, so we can detect the need to save later.
+	o.curRefresh = o.Token.RefreshToken
+
+	return fn, f.Close()
+}
+
+func load() (*openidCfg, error) {
+	dataDir := cfgpath.GetDataPath(cliApp.Name)
+	os.MkdirAll(dataDir, 0755)
+	fn := filepath.Join(dataDir, "openid.cfg")
+
+	buf, err := ioutil.ReadFile(fn)
+	if err != nil {
+		return nil, err
+	}
+
+	_, x, err := network.Unmarshal(buf, cothority.Suite)
+	if err != nil {
+		return nil, err
+	}
+	ocfg, ok := (x).(*openidCfg)
+	if !ok {
+		return nil, errors.New("wrong type")
+	}
+
+	// Mark this token as expired, because the round-trip thru save/load has caused us to lose
+	// the raw field, which is not exported. So force a refresh on next use.
+	ocfg.Token.Expiry = time.Unix(0, 0)
+
+	return ocfg, err
+}
+
+func (o *openidCfg) getSigners(cl *eventlog.Client) ([]darc.Signer, error) {
+	cfg := &o.Config
+	ts := cfg.TokenSource(context.Background(), &o.Token)
+
+	r := cl.ByzCoin.Roster
+	n := len(r.List)
+	T := threshold(n)
+
+	// The callback from darc.Sign where we need to go contact the Authentication Proxies.
+	cb := func(msg []byte) ([]byte, error) {
+		tok, err := ts.Token()
+		if err != nil {
+			return nil, err
+		}
+		// After calling Token(), o.Token might have been updated, using up the old refresh
+		// code. Call save, which will detect if the save is really needed or not.
+		o.Token = *tok
+		o.save()
+
+		rawIDToken, ok := tok.Extra("id_token").(string)
+		if !ok {
+			return nil, errors.New("no id_token in token response")
+		}
+
+		// Make the random shares for the signature.
+		rPri := share.NewPriPoly(cothority.Suite, T, nil, cothority.Suite.RandomStream())
+		rShares := rPri.Shares(n)
+		rPub := rPri.Commit(nil)
+		_, rPubCommits := rPub.Info()
+
+		// Make a shuffled list of server id's to contact.
+		shuffled := rand.Perm(len(r.List))
+		client := onet.NewClient(cothority.Suite, authprox.ServiceName)
+
+		// Connect to servers, get sigs, stop when we achieve the threshold.
+		// TODO: This could be in parallel.
+		var partials []*share.PriShare
+		for _, idx := range shuffled {
+			s := r.List[idx]
+
+			if len(partials) >= T {
+				// Got enough, all done.
+				break
```

Comment:
breaking here means we're trusting all the proxies, which I don't think is in the threat model. we should collect all the partial signatures and then remove any that are not valid (not sure how, some kind of NIZK proof?). does DSS have this kind of functionality?

Reply:
DSS callers are expected to check the signature on the PartialSignature, which we are not yet doing. But that does not address the case where a good signature on an incorrect PartialSignature makes the entire reconstruction fail.

I will put a TODO here for now, and discuss with cryptographers to see what they say.